elastic/apm-server
# beater/interceptors: add Timeout interceptor

## Motivation/summary

Introduce an interceptor that intercepts `context.Canceled` and `context.DeadlineExceeded` errors, and sets the gRPC status code to DeadlineExceeded and message to "request timed out". Also, increment the request timeout counter.

## Checklist


## How to test these changes

1. Using a gRPC-based client (OpenTelemetry or Jaeger), send some spans to the server with the timeout very low
2. Check server request logs: `grpc.response.status_code` should be "DeadlineExceeded", and `error.message` should be "request timed out".

I'm not sure what the best way to force a request timeout is. Using opentelemetry-go, I resorted to wrapping [`SpanExporter.ExportSpans`](https://github.com/open-telemetry/opentelemetry-go/blob/6db20e008dc272efaa29bb87d5d8341ad7c06da0/sdk/trace/span_exporter.go#L33), injecting a timeout there.

## Related issues

Closes #5049
